### PR TITLE
perf(es/minifier): refactor base54

### PR DESCRIPTION
**Pull request by magic-akari**
**Description:**

**BREAKING CHANGE:**

**Related issue (if exists):**

### CODE REVIEW

File: crates/swc_ecma_minifier/tests/mangle/reuse/3/output.js [hunk 1]
```diff
@@ -1267,9 +1267,9 @@ class aa extends Array {
         };
     }
 }
-class ba {
+class ab {
```
Comment (kdy1) [inline crates/swc_ecma_minifier/tests/mangle/reuse/3/output.js:1270]: Is this intentional?
Just curious, as the previous impl may be wrong.

Comment (magic-akari) [reply crates/swc_ecma_minifier/tests/mangle/reuse/3/output.js:1270]: Yes.
It generate 
a, b, c ... z, A, B, C ... Z, $, _, 0,1, ... 9,
and
aa, ab, ac ...

In previous implementations, it seems that numbers were never used at the end.

**Comment by magic-akari** [comment]
Some of the usize values exceed the wasm limit.
Does this mean that we will never reach those values?

**Comment by kdy1** [comment]
I think so?
I'm sure, the process will be killed before we reach it.

**Comment by kdy1** [comment]
Oh then we can use smaller integer type

**Comment by magic-akari** [comment]
Should I remove those values for WASM, or remove them for all targets?
The max safe value for WASM is 4294967295, the max u32.

**Comment by kdy1** [comment]
I think both are fine, but I prefer removing from both target.

File: crates/swc_ecma_minifier/src/util/base54.rs [hunk 2]
```diff
@@ -1,46 +1,198 @@
...
+        x <<= 6;
+        x += reverse_base54(c);
+    }
+
+    ret += x;
+
+    ret
+}
+
+fn reverse_base54(c: &u8) -> usize {
```
Comment (kdy1) [inline crates/swc_ecma_minifier/src/util/base54.rs:67]: I think accepting `u8` will be better for performance.

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thank you!

I think this is a quite novel idea.

File: crates/swc_ecma_minifier/src/util/base54.rs (hunk 2 above)
Comment (magic-akari) [reply crates/swc_ecma_minifier/src/util/base54.rs]: The `decode` and `reverse_base54` functions are currently not used anywhere except for testing.

Comment (kdy1) [reply crates/swc_ecma_minifier/src/util/base54.rs]: I see, then it's fine.
Thanks for the explanation

File: crates/swc_ecma_minifier/tests/mangle/reuse/3/output.js (hunk 1 above)
Comment (magic-akari) [reply crates/swc_ecma_minifier/tests/mangle/reuse/3/output.js:1270]: Oh, numeric endings are possible in previous implementation.
Now it is easier to implement `encode` and `decode` after changing the rules.

File: crates/swc_ecma_minifier/src/util/base54.rs [hunk 3]
```diff
@@ -1,46 +1,209 @@
...
+    98312762,   // false
+    321674951,  // super
+    335053132,  // throw
+    385347706,  // while
+    4211585658, // delete
+];
+
+// u32::MAX is 4294967295
+
+// const RESERVED_U64: [usize; 13] = [
```
Comment (Brooooooklyn) [inline crates/swc_ecma_minifier/src/util/base54.rs:117]: Are these commented codes necessary? If not, please delete them.

File: crates/swc_ecma_minifier/src/pass/compute_char_freq.rs [hunk 4]
```diff
@@ -19,7 +19,7 @@ pub(crate) struct CharFreqInfo {
 pub(crate) fn compute_char_freq(_m: &Module) -> CharFreqInfo {
```
Comment (magic-akari) [inline crates/swc_ecma_minifier/src/pass/compute_char_freq.rs:19]: @kdy1 
CharFreqInfo change the `BASE54_DEFAULT_CHARS` order which will break the `is_reserved` function

Comment (kdy1) [reply crates/swc_ecma_minifier/src/pass/compute_char_freq.rs:19]: Ah...
Then I think it's fine to remove `CharFreqInfo` for now, as it does not have enough logic to be useful.

Comment (magic-akari) [reply crates/swc_ecma_minifier/src/pass/compute_char_freq.rs:19]: Would using CharFreqInfo achieve better gzip results?
If so, I'd prefer to keep it. 
I can generate the reserved list at runtime.

Comment (kdy1) [reply crates/swc_ecma_minifier/src/pass/compute_char_freq.rs:19]: Not yet at the moment, seems like we need full codegen or smart visitor to make the compression better.

**Review by kdy1 — APPROVED** [review]
LGTM. Thank you!

---

swc-bump:
 - swc_ecma_minifier

**Pull request merged**